src/components/producerItems/components/ProductEditModal.js

```diff
@@ -0,0 +1,114 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import { Form, Input, Modal } from 'antd';
+
+import {
+    DEFAULT_FIRST_FIELD,
+    EDIT_FORM_TITLE,
+    EDITABLE_FIELDS,
+    FORM_ITEM_LAYOUT,
+    SAVE_PRODUCT_DETAILS
+} from '../constants';
+
+const FormItem = Form.Item;
+
+/**
+ * Helper function used to generate input fields for `ProductEditForm`.
+ * It makes sure that the title/name field is at the beginning of the edit form
+ *
+ * @function
+
+ * @param {Function} decorator - sets the appropriate rules and defaults on the
+ * input field
+ * @param {Object} productToEdit - the product details to edit
+ * @return {Array} - a list of field inputs
+ */
+function generateProductEditForm(decorator, productToEdit) {
+    const productFields = Object.entries(productToEdit);
+
+    return productFields.reduce((total, item) => {
+        const [field, initialValue] = item;
+        const editableField = EDITABLE_FIELDS[field];
+
+        if (editableField) {
+            const { label, rules } = editableField;
+            const formItem = (
+                <FormItem
+                    {...FORM_ITEM_LAYOUT}
+                    key={field}
+                    label={label}
+                >
+                    { decorator(field, { initialValue, rules })(<Input />) }
+                </FormItem>
+            );
+
+            field === DEFAULT_FIRST_FIELD ? total.unshift(formItem) : total.push(formItem);
+        }
+
+        return total;
+    }, []);
+}
+
+/**
+ * React component used to render the product eidtable fields
+ */
+class ProductEditForm extends React.Component {
+    state = { formItems: [] };
+
+    /**
+     * Validates the the form rules have all passed then sends a request to update
+     * the product details
+     *
+     * @function
+     * @return {Void} - void
+     */
+    updateProductInfo = () => {
+        const { form, productToEdit, updateProduct } = this.props;
+
+        form.validateFields((error, productValues ) => error ?
+            error :
+            updateProduct({ ...productToEdit, ...productValues })
+        );
+    }
+
+    static getDerivedStateFromProps(props) {
+        const { productToEdit, form: { getFieldDecorator } } = props;
+        const formItems = generateProductEditForm(getFieldDecorator, productToEdit);
+
+        return { formItems };
+    }
```

Comment:
why did you use static here?
and what is the method doing?

Reply:
what is going on here?

Reply:
[`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) is a new react life cycle introduce in react v16. It is the safest way to update your component state from the properties that it will receive. It is [`static`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static) because it does not need access to the component class to work, meaning that you don't have access to `this` within the function